src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -39,6 +39,167 @@ std::vector<FixedPoint> VectorDoubleToVectorFixedPoint(const std::vector<double>
 std::vector<double> VectorFixedPointToVectorDouble(
     const std::vector<FixedPoint> &vector_fp);
 
+/// \class ResourceSet
+/// \brief Encapsulates and operates on a set of resources, including CPUs,
+/// GPUs, and custom labels.
+class ResourceSet {
+ public:
+  static std::shared_ptr<ResourceSet> Nil() {
+    static auto nil = std::make_shared<ResourceSet>();
+    return nil;
+  }
+
+  /// \brief empty ResourceSet constructor.
+  ResourceSet();
+
+  /// \brief Constructs ResourceSet from the specified resource map.
+  ResourceSet(const std::unordered_map<std::string, FixedPoint> &resource_map);
+
+  /// \brief Constructs ResourceSet from the specified resource map.
+  ResourceSet(const std::unordered_map<std::string, double> &resource_map);
+
+  /// \brief Constructs ResourceSet from two equal-length vectors with label and capacity
+  /// specification.
+  ResourceSet(const std::vector<std::string> &resource_labels,
+              const std::vector<double> resource_capacity);
+
+  /// \brief Empty ResourceSet destructor.
+  ~ResourceSet() = default;
+
+  /// \brief Test equality with the other specified ResourceSet object.
+  ///
+  /// \param rhs: Right-hand side object for equality comparison.
+  /// \return True if objects are equal, False otherwise.
+  bool operator==(const ResourceSet &rhs) const;
+
+  /// \brief Test equality with the other specified ResourceSet object.
+  ///
+  /// \param other: Right-hand side object for equality comparison.
+  /// \return True if objects are equal, False otherwise.
+  bool IsEqual(const ResourceSet &other) const;
```

Comment:
Why need both `operator==` and `IsEqual`. This may cause confusion.

Reply:
This is copied from the old data structure, let's keep it as original in this PR to avoid potential bugs. We can open another PR to do it after this PR is merged.